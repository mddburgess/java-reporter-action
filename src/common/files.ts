import * as glob from '@actions/glob';

export async function findFiles(searchPaths: string[]) {
    const globber = await glob.create(searchPaths.join('\n'));
    return await globber.glob();
}

export async function findFile(searchPath: string) {
    const paths = await findFiles([searchPath]);
    return paths[0];
}

export async function findRelativePath(searchPath: string) {
    const globber = await glob.create(searchPath);
    const searchRoot = globber.getSearchPaths()[0];
    const absolutePaths = await globber.glob();
    return absolutePaths[0] && absolutePaths[0].slice(searchRoot.length + 1);
}
