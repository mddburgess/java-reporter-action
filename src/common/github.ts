export interface Annotation {
    path: string,
    start_line: number,
    end_line: number,
    start_column?: number,
    end_column?: number,
    annotation_level: AnnotationLevel,
    title?: string,
    message: string,
    raw_details?: string
}

export type AnnotationLevel =
    | 'notice'
    | 'warning'
    | 'failure';
