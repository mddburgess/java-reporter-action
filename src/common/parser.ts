import * as core from "@actions/core";
import * as fs from "fs";
import { Saxophone } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";

export default abstract class ReportParser<T> {
  private readonly context: string[] = ["root"];
  private readonly parser = new Saxophone()
    .on("tagOpen", (tag) => this.handleTagOpen(tag))
    .on("tagClose", (tag) => this.handleTagClose(tag))
    .on("text", (tag) => this.handleText(tag))
    .on("cdata", (tag) => this.handleText(tag));

  protected constructor(protected readonly report: T, private readonly reportPath: string) {}

  read() {
    try {
      const xml = fs.readFileSync(this.reportPath, { encoding: "utf-8" });
      this.parser.parse(xml);
      return this.report;
    } catch (error) {
      core.warning(`Failed to read report: ${this.reportPath}`);
      return undefined;
    }
  }

  protected getContext() {
    return this.context[0];
  }

  private handleTagOpen(tag: TagOpenNode) {
    this.onTagOpen(tag);
    if (tag.isSelfClosing) {
      this.onTagClose(tag);
    } else {
      this.context.unshift(tag.name);
    }
  }

  private handleTagClose(tag: TagCloseNode) {
    this.context.shift();
    this.onTagClose(tag);
  }

  private handleText(tag: TextNode | CDATANode) {
    this.onText(tag);
  }

  protected abstract onTagOpen(tag: TagOpenNode): void;
  protected abstract onTagClose(tag: TagCloseNode): void;
  protected abstract onText(tag: TextNode | CDATANode): void;
}
