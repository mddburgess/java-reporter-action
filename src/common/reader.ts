import * as core from '@actions/core';
import * as fs from 'fs';
import {Saxophone} from 'saxophone-ts';
import {CDATANode, TagCloseNode, TagOpenNode, TextNode} from 'saxophone-ts/dist/types/src/static/nodes';

abstract class ReportReader<T> {

    private readonly parser = new Saxophone()
        .on('tagOpen', tag => this.handleTagOpen(tag))
        .on('tagClose', tag => this.handleTagClose(tag))
        .on('text', tag => this.handleText(tag))
        .on('cdata', tag => this.handleText(tag));
    private readonly context: string[] = ['root'];
    protected report?: T;

    public readReport(reportPath: string): T | undefined {
        try {
            const xml = fs.readFileSync(reportPath, {encoding: 'utf-8'});
            this.parser.parse(xml);
            return this.report;
        } catch (error) {
            core.warning(`Failed to read report: ${reportPath}`);
            console.log(error);
            return undefined;
        }
    }

    protected getContext(): string {
        return this.context[0];
    }

    private handleTagOpen(tag: TagOpenNode): void {
        this.onTagOpen(tag);
        if (tag.isSelfClosing) {
            this.onTagClose(tag);
        } else {
            this.context.unshift(tag.name);
        }
    }

    private handleTagClose(tag: TagCloseNode): void {
        this.context.shift();
        this.onTagClose(tag);
    }

    private handleText(tag: TextNode | CDATANode): void {
        this.onText(tag);
    }

    protected abstract onTagOpen(tag: TagOpenNode): void;
    protected abstract onTagClose(tag: TagCloseNode): void;
    protected abstract onText(tag: TextNode | CDATANode): void;
}

export default ReportReader;
