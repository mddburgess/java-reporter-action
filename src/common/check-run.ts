import * as core from '@actions/core';
import * as github from '@actions/github';
import {Annotation} from './github';

class CheckRun {

    private readonly token;
    private readonly octokit;

    private readonly name;
    private checkRunId?: number;

    constructor(name: string) {
        this.token = core.getInput('github-token', {required: true});
        this.octokit = github.getOctokit(this.token);
        this.name = name;
    }

    public async conclude(request: CheckRunRequest) {
        if (request.output) {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion,
                output: {
                    title: request.output.title,
                    summary: request.output.summary
                }
            });
            await this.annotate(request.output);
        } else {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion
            });
        }
    }

    private async annotate(output: CheckOutput) {
        if (!output.annotations || output.annotations.length === 0) {
            return;
        }
        for (let i = 0; i < output.annotations.length; i += 50) {
            await this.updateCheckRun({
                output: {
                    title: output.title,
                    summary: output.summary,
                    annotations: output.annotations.slice(i, i + 50)
                }
            });
        }
    }

    private async createOrUpdateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            await this.createCheckRun(request);
        } else {
            await this.updateCheckRun(request);
        }
    }

    private async createCheckRun(request: CheckRunRequest) {
        core.debug(`Creating ${this.name} check run...`);
        const response = await this.octokit.checks.create({
            ...github.context.repo,
            name: this.name,
            head_sha: this.resolveHeadSha(),
            ...request
        });
        this.checkRunId = response.data.id;
    }

    private async updateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            throw Error('Cannot update a check run before creating it');
        }

        core.debug(`Updating ${this.name} check run...`);
        await this.octokit.checks.update({
            ...github.context.repo,
            check_run_id: this.checkRunId,
            ...request
        });
    }

    private resolveHeadSha() {
        return github.context.payload.pull_request
            ? github.context.payload.pull_request.head.sha
            : github.context.sha;
    }
}

interface CheckRunRequest {
    status?: CheckStatus,
    conclusion?: CheckConclusion,
    output?: CheckOutput
}

type CheckStatus =
    | 'queued'
    | 'in_progress'
    | 'completed';

type CheckConclusion =
    | 'success'
    | 'failure'
    | 'neutral'
    | 'cancelled'
    | 'skipped'
    | 'timed_out'
    | 'action_required';

interface CheckOutput {
    title: string,
    summary: string,
    text?: string,
    annotations?: Annotation[];
}

export default CheckRun;
