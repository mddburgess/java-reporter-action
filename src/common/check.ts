import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import path from 'path';
import {getInput} from '@actions/core';

abstract class Check {

    private static readonly workspacePath = process.env.GITHUB_WORKSPACE || '';

    private readonly reportType: string;
    private readonly checkCondition: CheckCondition;
    private checkRunId?: number;

    protected constructor(reportType: string) {
        this.reportType = reportType;
        this.checkCondition = this.resolveCheckCondition();
    }

    private resolveCheckCondition() {
        const condition = getInput(this.reportType);
        switch (condition) {
            case 'required':
            case 'expected':
            case 'disabled':
                return CheckCondition[condition];
            case undefined:
                return CheckCondition.autodetect;
            default:
                core.warning(`Input '${this.reportType}' is invalid - must be one of ['required','expected','disabled']. Defaulting to 'expected'.`);
                return CheckCondition.expected;
        }
    }

    public async run() {
        const reportPaths = await this.findReports();

        const token = core.getInput('github-token', {required: true});
        const octokit = github.getOctokit(token);

        if (reportPaths.length > 0 || this.checkCondition >= CheckCondition.expected) {
            const response = await octokit.checks.create({
                ...github.context.repo,
                name: this.reportType,
                head_sha: github.context.payload.pull_request
                    ? github.context.payload.pull_request.head.sha
                    : github.context.sha,
                status: 'completed',
                conclusion: reportPaths.length ? 'success' :
                    this.checkCondition === CheckCondition.required ? 'failure' : 'skipped'
            });
            this.checkRunId = response.data.id;
        }

        core.info(`${this.reportType} check finished.`);
    }

    private async findReports() {
        core.startGroup(`Searching for ${this.reportType} reports...`);

        const searchPaths = this.reportSearchPaths();
        searchPaths.forEach(searchPath => core.info(searchPath));
        core.endGroup();

        const globber = await glob.create(searchPaths.join('\n'));
        const reportPaths = await globber.glob();

        core.startGroup(`Found ${reportPaths.length} ${this.reportType} reports.`);
        reportPaths.forEach(reportPath => core.info(path.relative(Check.workspacePath, reportPath)));
        core.endGroup();

        return reportPaths;
    }

    protected abstract reportSearchPaths(): string[];
}

enum CheckCondition {
    disabled,
    autodetect,
    expected,
    required
}

export default Check;
