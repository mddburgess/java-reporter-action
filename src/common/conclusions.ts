import { AnnotationSummary } from "../github/check-run";
import { CheckConclusion } from "../github/types";

const failureOnWarnings = (result: AnnotationSummary): CheckConclusion => {
  if (result.failures + result.warnings > 0) {
    return "failure";
  }
  return "success";
};

const neutralOnWarnings = (result: AnnotationSummary): CheckConclusion => {
  if (result.failures > 0) {
    return "failure";
  }
  if (result.warnings > 0) {
    return "neutral";
  }
  return "success";
}

export {
  failureOnWarnings,
  neutralOnWarnings
};
