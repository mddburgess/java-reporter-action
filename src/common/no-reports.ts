import CheckResult from "../check/result";
import { CheckAnnotation, CheckConclusion } from "../github/types";
import { RunCondition } from "../check";

export default class NoReportsResult extends CheckResult {
  constructor(
    private readonly friendlyName: string,
    private readonly runCondition: RunCondition,
    private readonly searchPaths: string[]
  ) {
    super();
  }

  shouldCompleteCheck(): boolean {
    return this.runCondition >= RunCondition.expected;
  }

  get conclusion(): CheckConclusion {
    return this.runCondition === RunCondition.required ? "failure" : "skipped";
  }

  get title(): string {
    return "No reports found";
  }

  get summary(): string {
    const runConditionName = RunCondition.required ? "required" : "expected";
    return `The ${this.friendlyName} check is ${runConditionName}, but no ${this.friendlyName} reports were found.`;
  }

  get text(): string | undefined {
    return ["### Search paths", "```sh", ...this.searchPaths, "```"].join("\n");
  }

  get annotations(): CheckAnnotation[] | undefined {
    return undefined;
  }
}
