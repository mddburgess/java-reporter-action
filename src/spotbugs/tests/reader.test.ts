import {findFile} from '../../common/files';
import SpotbugsReportReader from '../reader';

describe('SpotBugs report reader', () => {

    it('can read a SpotBugs report', async () => {
        const reportPath = await findFile('**/spotbugsXml.xml');
        const report = new SpotbugsReportReader().readReport(reportPath);
        expect(report).toMatchSnapshot();
    });
});
