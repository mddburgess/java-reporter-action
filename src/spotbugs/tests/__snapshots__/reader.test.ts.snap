// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotBugs report reader can read a SpotBugs report 1`] = `
Object {
  "bugs": Array [
    Object {
      "category": "CORRECTNESS",
      "endLine": 142,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Impossible downcast of toArray() result to Integer[] in org.example.cpd.ErrorProne.classCastExceptionWithToArray()",
      "priority": 1,
      "shortMessage": "Impossible downcast of toArray() result",
      "startLine": 142,
    },
    Object {
      "category": "STYLE",
      "endLine": 142,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Dead store to array in org.example.cpd.ErrorProne.classCastExceptionWithToArray()",
      "priority": 2,
      "shortMessage": "Dead store to local variable",
      "startLine": 142,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 116,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "BigDecimal constructed from 1.123 in org.example.cpd.ErrorProne.avoidDecimalLiteralsInBigDecimalConstructor()",
      "priority": 2,
      "shortMessage": "BigDecimal constructed from double that isn't represented precisely",
      "startLine": 116,
    },
    Object {
      "category": "PERFORMANCE",
      "endLine": 37,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Primitive boxed just to call toString in org.example.cpd.ErrorProne.unnecessaryConversionTemporary(int)",
      "priority": 2,
      "shortMessage": "Method allocates a boxed primitive just to call toString",
      "startLine": 37,
    },
    Object {
      "category": "PERFORMANCE",
      "endLine": 37,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "org.example.cpd.ErrorProne.unnecessaryConversionTemporary(int) invokes inefficient new Integer(int) constructor; use Integer.valueOf(int) instead",
      "priority": 2,
      "shortMessage": "Method invokes inefficient Number constructor; use static valueOf instead",
      "startLine": 37,
    },
    Object {
      "category": "MT_CORRECTNESS",
      "endLine": 89,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Empty synchronized block in org.example.cpd.ErrorProne.emptySynchronizedBlock()",
      "priority": 2,
      "shortMessage": "Empty synchronized block",
      "startLine": 89,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 100,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "There is an apparent infinite loop in org.example.cpd.ErrorProne.jumbledIncrementer()",
      "priority": 1,
      "shortMessage": "An apparent infinite loop",
      "startLine": 100,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 128,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Null pointer dereference of string in org.example.cpd.ErrorProne.brokenNullCheck(String)",
      "priority": 1,
      "shortMessage": "Null pointer dereference",
      "startLine": 128,
    },
    Object {
      "category": "STYLE",
      "endLine": 128,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Load of known null value in org.example.cpd.ErrorProne.brokenNullCheck(String)",
      "priority": 2,
      "shortMessage": "Load of known null value",
      "startLine": 128,
    },
    Object {
      "category": "EXPERIMENTAL",
      "endLine": 137,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "org.example.cpd.ErrorProne.checkSkipResult() may fail to clean up java.io.InputStream",
      "priority": 2,
      "shortMessage": "Method may fail to clean up stream or resource",
      "startLine": 137,
    },
    Object {
      "category": "BAD_PRACTICE",
      "endLine": 135,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "org.example.cpd.ErrorProne.checkSkipResult() may fail to close stream",
      "priority": 2,
      "shortMessage": "Method may fail to close stream",
      "startLine": 135,
    },
    Object {
      "category": "STYLE",
      "endLine": 100,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Complicated, subtle or wrong increment in for-loop org.example.cpd.ErrorProne.jumbledIncrementer()",
      "priority": 2,
      "shortMessage": "Complicated, subtle or wrong increment in for-loop",
      "startLine": 100,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 14,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Nullcheck of string at line 14 of value previously dereferenced in org.example.cpd.ErrorProne.misplacedNullCheck(String)",
      "priority": 2,
      "shortMessage": "Nullcheck of value previously dereferenced",
      "startLine": 14,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 46,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Return value of java.math.BigDecimal.add(BigDecimal) ignored in org.example.cpd.ErrorProne.uselessOperationOnImmutable()",
      "priority": 1,
      "shortMessage": "Method ignores return value",
      "startLine": 46,
    },
    Object {
      "category": "BAD_PRACTICE",
      "endLine": 136,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "org.example.cpd.ErrorProne.checkSkipResult() ignores result of java.io.FileInputStream.skip(long)",
      "priority": 2,
      "shortMessage": "Method ignores results of InputStream.skip()",
      "startLine": 136,
    },
    Object {
      "category": "STYLE",
      "endLine": 140,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Useless object stored in variable collection of method org.example.cpd.ErrorProne.classCastExceptionWithToArray()",
      "priority": 2,
      "shortMessage": "Useless object created",
      "startLine": 140,
    },
    Object {
      "category": "STYLE",
      "endLine": 113,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Method org.example.cpd.ErrorProne.avoidBranchingStatementAsLastInLoop() seems to be useless",
      "priority": 2,
      "shortMessage": "Useless non-empty void method",
      "startLine": 113,
    },
    Object {
      "category": "STYLE",
      "endLine": 143,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Method org.example.cpd.ErrorProne.classCastExceptionWithToArray() seems to be useless",
      "priority": 2,
      "shortMessage": "Useless non-empty void method",
      "startLine": 143,
    },
    Object {
      "category": "STYLE",
      "endLine": 53,
      "filePath": "org/example/cpd/ErrorProne.java",
      "longMessage": "Method org.example.cpd.ErrorProne.dontUseFloatTypeForLoopIndices() seems to be useless",
      "priority": 2,
      "shortMessage": "Useless non-empty void method",
      "startLine": 53,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 16,
      "filePath": "org/example/pmd/BestPractices.java",
      "longMessage": "Null pointer dereference of conn in org.example.pmd.BestPractices.checkResultSet()",
      "priority": 1,
      "shortMessage": "Null pointer dereference",
      "startLine": 16,
    },
    Object {
      "category": "EXPERIMENTAL",
      "endLine": 19,
      "filePath": "org/example/pmd/BestPractices.java",
      "longMessage": "org.example.pmd.BestPractices.checkResultSet() may fail to clean up java.sql.ResultSet",
      "priority": 2,
      "shortMessage": "Method may fail to clean up stream or resource",
      "startLine": 19,
    },
    Object {
      "category": "EXPERIMENTAL",
      "endLine": 19,
      "filePath": "org/example/pmd/BestPractices.java",
      "longMessage": "org.example.pmd.BestPractices.checkResultSet() may fail to clean up java.sql.Statement",
      "priority": 2,
      "shortMessage": "Method may fail to clean up stream or resource",
      "startLine": 19,
    },
    Object {
      "category": "BAD_PRACTICE",
      "endLine": 16,
      "filePath": "org/example/pmd/BestPractices.java",
      "longMessage": "org.example.pmd.BestPractices.checkResultSet() may fail to close Statement",
      "priority": 2,
      "shortMessage": "Method may fail to close database resource",
      "startLine": 16,
    },
    Object {
      "category": "STYLE",
      "endLine": 11,
      "filePath": "org/example/pmd/BestPractices.java",
      "longMessage": "Unread public/protected field: org.example.pmd.BestPractices.ipAddress",
      "priority": 2,
      "shortMessage": "Unread public/protected field",
      "startLine": 11,
    },
    Object {
      "category": "PERFORMANCE",
      "endLine": NaN,
      "filePath": "org/example/pmd/BestPractices.java",
      "longMessage": "Unused field: org.example.pmd.BestPractices.unusedPrivateField",
      "priority": 2,
      "shortMessage": "Unused field",
      "startLine": NaN,
    },
    Object {
      "category": "STYLE",
      "endLine": 41,
      "filePath": "org/example/pmd/CodeStyle.java",
      "longMessage": "Dead store to instance in org.example.pmd.CodeStyle.uselessQualifiedThis()",
      "priority": 2,
      "shortMessage": "Dead store to local variable",
      "startLine": 41,
    },
    Object {
      "category": "STYLE",
      "endLine": 8,
      "filePath": "org/example/pmd/Design.java",
      "longMessage": "Dead store to x in org.example.pmd.Design.collapsibleIfStatements(boolean, boolean)",
      "priority": 2,
      "shortMessage": "Dead store to local variable",
      "startLine": 8,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 51,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Impossible downcast of toArray() result to Integer[] in org.example.pmd.ErrorProne.classCastExceptionWithToArray()",
      "priority": 1,
      "shortMessage": "Impossible downcast of toArray() result",
      "startLine": 51,
    },
    Object {
      "category": "STYLE",
      "endLine": 51,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Dead store to array in org.example.pmd.ErrorProne.classCastExceptionWithToArray()",
      "priority": 2,
      "shortMessage": "Dead store to local variable",
      "startLine": 51,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 25,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "BigDecimal constructed from 1.123 in org.example.pmd.ErrorProne.avoidDecimalLiteralsInBigDecimalConstructor()",
      "priority": 2,
      "shortMessage": "BigDecimal constructed from double that isn't represented precisely",
      "startLine": 25,
    },
    Object {
      "category": "PERFORMANCE",
      "endLine": 135,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Primitive boxed just to call toString in org.example.pmd.ErrorProne.unnecessaryConversionTemporary(int)",
      "priority": 2,
      "shortMessage": "Method allocates a boxed primitive just to call toString",
      "startLine": 135,
    },
    Object {
      "category": "PERFORMANCE",
      "endLine": 135,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "org.example.pmd.ErrorProne.unnecessaryConversionTemporary(int) invokes inefficient new Integer(int) constructor; use Integer.valueOf(int) instead",
      "priority": 2,
      "shortMessage": "Method invokes inefficient Number constructor; use static valueOf instead",
      "startLine": 135,
    },
    Object {
      "category": "MT_CORRECTNESS",
      "endLine": 94,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Empty synchronized block in org.example.pmd.ErrorProne.emptySynchronizedBlock()",
      "priority": 2,
      "shortMessage": "Empty synchronized block",
      "startLine": 94,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 105,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "There is an apparent infinite loop in org.example.pmd.ErrorProne.jumbledIncrementer()",
      "priority": 1,
      "shortMessage": "An apparent infinite loop",
      "startLine": 105,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 37,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Null pointer dereference of string in org.example.pmd.ErrorProne.brokenNullCheck(String)",
      "priority": 1,
      "shortMessage": "Null pointer dereference",
      "startLine": 37,
    },
    Object {
      "category": "STYLE",
      "endLine": 37,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Load of known null value in org.example.pmd.ErrorProne.brokenNullCheck(String)",
      "priority": 2,
      "shortMessage": "Load of known null value",
      "startLine": 37,
    },
    Object {
      "category": "EXPERIMENTAL",
      "endLine": 46,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "org.example.pmd.ErrorProne.checkSkipResult() may fail to clean up java.io.InputStream",
      "priority": 2,
      "shortMessage": "Method may fail to clean up stream or resource",
      "startLine": 46,
    },
    Object {
      "category": "BAD_PRACTICE",
      "endLine": 44,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "org.example.pmd.ErrorProne.checkSkipResult() may fail to close stream",
      "priority": 2,
      "shortMessage": "Method may fail to close stream",
      "startLine": 44,
    },
    Object {
      "category": "STYLE",
      "endLine": 105,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Complicated, subtle or wrong increment in for-loop org.example.pmd.ErrorProne.jumbledIncrementer()",
      "priority": 2,
      "shortMessage": "Complicated, subtle or wrong increment in for-loop",
      "startLine": 105,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 112,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Nullcheck of string at line 112 of value previously dereferenced in org.example.pmd.ErrorProne.misplacedNullCheck(String)",
      "priority": 2,
      "shortMessage": "Nullcheck of value previously dereferenced",
      "startLine": 112,
    },
    Object {
      "category": "CORRECTNESS",
      "endLine": 144,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Return value of java.math.BigDecimal.add(BigDecimal) ignored in org.example.pmd.ErrorProne.uselessOperationOnImmutable()",
      "priority": 1,
      "shortMessage": "Method ignores return value",
      "startLine": 144,
    },
    Object {
      "category": "BAD_PRACTICE",
      "endLine": 45,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "org.example.pmd.ErrorProne.checkSkipResult() ignores result of java.io.FileInputStream.skip(long)",
      "priority": 2,
      "shortMessage": "Method ignores results of InputStream.skip()",
      "startLine": 45,
    },
    Object {
      "category": "STYLE",
      "endLine": 49,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Useless object stored in variable collection of method org.example.pmd.ErrorProne.classCastExceptionWithToArray()",
      "priority": 2,
      "shortMessage": "Useless object created",
      "startLine": 49,
    },
    Object {
      "category": "STYLE",
      "endLine": 22,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Method org.example.pmd.ErrorProne.avoidBranchingStatementAsLastInLoop() seems to be useless",
      "priority": 2,
      "shortMessage": "Useless non-empty void method",
      "startLine": 22,
    },
    Object {
      "category": "STYLE",
      "endLine": 52,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Method org.example.pmd.ErrorProne.classCastExceptionWithToArray() seems to be useless",
      "priority": 2,
      "shortMessage": "Useless non-empty void method",
      "startLine": 52,
    },
    Object {
      "category": "STYLE",
      "endLine": 58,
      "filePath": "org/example/pmd/ErrorProne.java",
      "longMessage": "Method org.example.pmd.ErrorProne.dontUseFloatTypeForLoopIndices() seems to be useless",
      "priority": 2,
      "shortMessage": "Useless non-empty void method",
      "startLine": 58,
    },
    Object {
      "category": "MT_CORRECTNESS",
      "endLine": 18,
      "filePath": "org/example/pmd/Multithreading.java",
      "longMessage": "Possible doublecheck on org.example.pmd.Multithreading.object in org.example.pmd.Multithreading.doubleCheckedLocking()",
      "priority": 2,
      "shortMessage": "Possible double check of field",
      "startLine": 16,
    },
    Object {
      "category": "STYLE",
      "endLine": 8,
      "filePath": "org/example/pmd/Multithreading.java",
      "longMessage": "Dead store to tg in org.example.pmd.Multithreading.avoidThreadGroup()",
      "priority": 1,
      "shortMessage": "Dead store to local variable",
      "startLine": 8,
    },
    Object {
      "category": "MT_CORRECTNESS",
      "endLine": 19,
      "filePath": "org/example/pmd/Multithreading.java",
      "longMessage": "Inconsistent synchronization of org.example.pmd.Multithreading.object; locked 40% of time",
      "priority": 2,
      "shortMessage": "Inconsistent synchronization",
      "startLine": 19,
    },
    Object {
      "category": "MT_CORRECTNESS",
      "endLine": 12,
      "filePath": "org/example/pmd/Multithreading.java",
      "longMessage": "org.example.pmd.Multithreading.dontCallThreadRun() explicitly invokes run on a thread (did you mean to start it instead?)",
      "priority": 2,
      "shortMessage": "Invokes run on a thread (did you mean to start it instead?)",
      "startLine": 12,
    },
    Object {
      "category": "PERFORMANCE",
      "endLine": 12,
      "filePath": "org/example/pmd/Performance.java",
      "longMessage": "org.example.pmd.Performance.booleanInstantiation() invokes inefficient Boolean constructor; use Boolean.valueOf(...) instead",
      "priority": 2,
      "shortMessage": "Method invokes inefficient Boolean constructor; use Boolean.valueOf(...) instead",
      "startLine": 12,
    },
  ],
  "categories": Map {
    "BAD_PRACTICE" => "Bad practice",
    "CORRECTNESS" => "Correctness",
    "PERFORMANCE" => "Performance",
    "STYLE" => "Dodgy code",
    "EXPERIMENTAL" => "Experimental",
    "MT_CORRECTNESS" => "Multithreaded correctness",
  },
}
`;
